concert/tests/integration/test_pcotimestamp_addon.py

```diff
@@ -0,0 +1,169 @@
+import shutil
+import tempfile
+from datetime import datetime
+import numpy as np
+
+from concert.quantities import q
+from concert.storage import DirectoryWalker
+from concert.tests import TestCase
+from concert.devices.cameras.pco import Timestamp
+from concert.devices.cameras.pco import Camera as PCOCamera
+from concert.devices.cameras.dummy import Camera as DummyCamera
+from concert.helpers import ImageWithMetadata
+from concert.experiments.synchrotron import Radiography
+from concert.devices.motors.dummy import LinearMotor
+from concert.devices.shutters.dummy import Shutter
+from concert.experiments.addons import PCOTimestampCheck, ImageWriter
+from concert.base import transition, Parameter, identity
+
+
+def _int_to_int_array(number, length):
+    array = np.zeros(length, dtype=np.uint16)
+    for i in range(length):
+        array[length-i-1] = number % 10
+        number //= 10
+    return array
+
+
+def add_binary_timestamp(img, time: datetime, number):
+    data = np.zeros(28, dtype=np.uint16)
+    data[:8] = _int_to_int_array(number, 8)
+    data[8:12] = _int_to_int_array(time.year, 4)
+    data[12:14] = _int_to_int_array(time.month, 2)
+    data[14:16] = _int_to_int_array(time.day, 2)
+    data[16:18] = _int_to_int_array(time.hour, 2)
+    data[18:20] = _int_to_int_array(time.minute, 2)
+    data[20:22] = _int_to_int_array(time.second, 2)
+    data[22:28] = _int_to_int_array(time.microsecond, 6)
+
+    for i in range(14):
+        d0 = data[i * 2] & 0xf
+        d1 = data[i * 2 + 1] & 0xf
+        img[0, i] = d0 << 4 | d1
+    return img
+
+
+class Camera(DummyCamera):
+    random_timestamp_numbers = Parameter()
+
+    async def __ainit__(self, background=None, simulate=True):
+        self.number = 1
+        self._random_numbers = None
+        self._timestamp_enabled = True
+        await DummyCamera.__ainit__(self, background, simulate)
+        await self.set_random_timestamp_numbers(False)
+
+    async def _get_random_timestamp_numbers(self):
+        return self._random_numbers
+
+    async def _set_random_timestamp_numbers(self, value: bool):
+        self._random_numbers = bool(value)
+
+    async def grab(self) -> ImageWithMetadata:
+        return await PCOCamera.grab(self)
+
+    @transition(target='recording')
+    async def _record_real(self):
+        self.number = 1
+
+    async def _grab_real(self):
+        img = await super()._grab_real()
+        time = datetime.now()
+        if not await self.get_random_timestamp_numbers():
+            img = add_binary_timestamp(img, time, self.number)
+        else:
+            img = add_binary_timestamp(img, time, np.random.randint(1, 1000))
+        self.number += 1
+        return img
+
+
+PCOCamera._grab_real = Camera._grab_real
```

Comment:
Or you could derive from the `DummyCamera` first _and_ `PCOCamera` second and override `_grab_real`. That way things like `start_recording` would be done on the dummy (left-right MRO). Basically, you can choose what parent gets called when, which is exactly what you need.